core/bcast/errors_rate.go

```diff
@@ -0,0 +1,95 @@
+// Copyright © 2022-2024 Obol Labs Inc. Licensed under the terms of a Business Source License 1.1
+
+package bcast
+
+import (
+	"sync/atomic"
+
+	"github.com/obolnetwork/charon/core"
+)
+
+const (
+	// RegistrationErrorsRateThreshold defines the threshold for registration errors rate.
+	RegistrationErrorsRateThreshold int32 = 70
+)
+
+// errorsRate object tracks registration errors rate within the last two epochs and updates metrics.
+type errorsRate struct {
+	pregenState     rateState
+	downstreamState rateState
+}
+
+// rateState represents single rate evaluation state.
+type rateState struct {
+	total    atomic.Int32
+	errors   atomic.Int32
+	prevRate atomic.Int32
+}
+
+// incrementTotal() increments the total counters.
+func (r *errorsRate) incrementTotal(duty core.Duty) {
+	if duty.Type != core.DutyBuilderRegistration {
+		return
+	}
+
+	if duty.Slot == 0 {
+		r.pregenState.total.Add(1)
+	} else {
+		r.downstreamState.total.Add(1)
+	}
+}
+
+// incrementErrors() increments the error counters.
+func (r *errorsRate) incrementErrors(duty core.Duty) {
+	if duty.Type != core.DutyBuilderRegistration {
+		return
+	}
+
+	if duty.Slot == 0 {
+		r.pregenState.errors.Add(1)
+	} else {
+		r.downstreamState.errors.Add(1)
+	}
+}
+
+// updateMetrics() updates metrics if a rate reached the threshold.
+func (r *errorsRate) updateMetrics() {
+	recastTotal.WithLabelValues(regSourcePregen).Add(float64(r.pregenState.total.Load()))
+	recastErrors.WithLabelValues(regSourcePregen).Add(float64(r.pregenState.errors.Load()))
+	recastTotal.WithLabelValues(regSourceDownstream).Add(float64(r.downstreamState.total.Load()))
+	recastErrors.WithLabelValues(regSourceDownstream).Add(float64(r.downstreamState.errors.Load()))
+
+	currentPregenRate := calculateRate(&r.pregenState)
+	currentDownstreamRate := calculateRate(&r.downstreamState)
+
+	// We want to be tolerant of the request failing in a single epoch, and warn when it happens for two consecutive epochs.
+	if r.pregenState.prevRate.Load() >= RegistrationErrorsRateThreshold && currentPregenRate >= RegistrationErrorsRateThreshold {
```

Comment:
The way it’s implemented, `recastErrorsRate` act as a boolean flag right?

E.g. if in the previous epoch we failed 10% of the requests, and 75% in the current epoch, we won’t trigger the warning.

Reply:
Taken offline for internal discussion.